pool/rate_limiting_average.go

```diff
@@ -72,18 +72,17 @@ func (p *ewmaRateLimiting) getNextTime(dur time.Duration) time.Duration {
 
 	p.ewma.Add(float64(dur))
 
-	adjustedRuntime := time.Duration(math.Ceil(p.ewma.Value())) / time.Duration(p.size)
-	runtimeOfTargetNumber := adjustedRuntime * time.Duration(p.target)
-
-	// if the expected runtime of the target number of tasks
-	// (adjisted for the size of the pool) is less than the stated
-	// period, return the difference between the total expected
-	// runtime of the tasks (adjusted) and the period, divided by
-	// the target number of tasks
-
+	// averageRuntime is the ema it takes to run one job
+	averageRuntime := time.Duration(math.Ceil(p.ewma.Value()))
+	// runtimeOfTargetNumber is how long running targetNum jobs will take
+	runtimeOfTargetNumber := averageRuntime * time.Duration(p.target) / time.Duration(p.size)
+
+	// if jobs will finish in the period, we should do the following:
+	// 1. figure out how much extra time there is in the period (p.period - runtimeOfTargetNumber)
+	// 2. divide this by the number of targets to find wait time per target
+	// 3. multiply by the number of workers, since each worker should wait (waitTimePerTarget * numWorkers)
 	if runtimeOfTargetNumber < p.period {
-		workerRestingTime := p.period - runtimeOfTargetNumber
-		return workerRestingTime / time.Duration(p.target)
+		return (p.period - runtimeOfTargetNumber) / time.Duration(p.target) * time.Duration(p.size)
```

Comment:
- feels like we shouldn't be able to get a duration from this division, but maybe I'm wrong and its rounding internally
- we divide by size on the top and multiply by it on the bottom of the final expression, which seems weird. 
- the fact that make sure that we're larger than the period on line 84, seems wrong, given that we ought to account for size here too.

Reply:
> feels like we shouldn't be able to get a duration from this division, but maybe I'm wrong and its rounding internally

There is no rounding. Everything is an integer in this expression. It's a duration (`p.period - runtimeOfTargetNumber`) divided by a constant and then multiplied by a constant. The reason for casting them to durations is to make Go's type system happy, since it won't let you multiply `time.Duration`s by integers (though `time.Duration`s are `int64`s internally).

- we divide by size on the top and multiply by it on the bottom of the final expression, which seems weird. 

We divide in order to find out the total remaining time to run. We then multiply the diff by size. This allows us to exit early if the expected runtime is less than 0. We could fold 78 into 85 and then return 0 if the result is negative.

- the fact that make sure that we're larger than the period on line 84, seems wrong, given that we ought to account for size here too.

See my reply to your previous bullet point.